File: problems/chapter11/tokuma/abc120_d.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct UnionFind
{
    vector<int> par, siz;

    // 初期化
    UnionFind(int n) : par(n, -1), siz(n, 1) {}

    // 根を求める
    // 経路圧縮
    // 根を求める
    int root(int x)
    {
        if (par[x] == -1)
        {
            return x; // x が根の場合は x を返す
        }
        else
        {
            return par[x] = root(par[x]);
        }
    }

    // x と y が同じグループに属するかどうか (根が一致するかどうか)
    bool issame(int x, int y)
    {
        return root(x) == root(y);
    }
    // x を含むグループと y を含むグループとを併合する
    bool unite(int x, int y)
    {
        // x, y をそれぞれ根まで移動する
        x = root(x);
        y = root(y);

        // すでに同じグループのときは何もしない
        if (x == y)
        {
            return false;
        }
        // union by size (y 側のサイズが小さくなるようにする)
        if (siz[x] < siz[y])
        {
            swap(x, y);
        }

        // y を x の子とする
        par[y] = x;
        siz[x] += siz[y];
        return true;
    }
    // xを含むグループのサイズ
    int size(int x)
    {
        return siz[root(x)];
    }
};

int main()
{

    int N, M;
    cin >> N >> M;

    vector<int> A(M), B(M);
    for (int i = 0; i < M; ++i)
    {
        cin >> A[i] >> B[i];
        --A[i];
        --B[i];
    }

    UnionFind uf(N);

    // 辺を追加していくことで、逆順に出力する
    // なにもないときはすべての島との組み合わせnC2だけコスト

    long long cur = N * (N - 1) / 2;
    vector<long long> res;

    // 辺0本からMまで勧めていく
    for (int i = 0; i < M; ++i)
    {
        res.push_back(cur);

        // どの島と島を結びつけているか
        int a = A[M - (i + 1)], b = B[M - (i + 1)];

        // もしすでに連結していれば、以下の部分を無視
        if (uf.issame(a, b))
        {
            continue;
        }

        //そうでなければ、連結成分それぞれのグループがつながるので、
        // N_a, N_bの掛け算だけコストが減る
        long long sa = uf.size(a), sb = uf.size(b);
        cur -= sa * sb;

        // 結合していく
        uf.unite(a, b);
    }

    // 逆順でかいておく
    reverse(res.begin(), res.end());
    for (int i = 0; i < res.size(); ++i)
        cout << res[i] << endl;
}

File: problems/chapter11/Ysi/abc120_d.py
class UnionFind():
    def __init__(self, n) -> None:
        self.par = [-1] * n
        self.siz = [1] * n

    def root(self, x):
        if self.par[x] == -1:
            return x
        else:
            self.par[x] = self.root(self.par[x])
            return self.par[x]

    def issame(self, x, y):
        return self.root(x) == self.root(y)

    def unite(self, x, y):
        x = self.root(x)
        y = self.root(y)

        if x == y:
            return False

        if self.siz[x] < self.siz[y]:
            x, y = y, x

        self.par[y] = x
        self.siz[x] += self.siz[y]
        return True

    def size(self, x):
        return self.siz[self.root(x)]

def main():
    N, M = map(int, input().split())
    edge = []
    for i in range(M):
        a,b = map(int, input().split())
        a -= 1
        b -= 1
        edge.append([a, b])
    edge.reverse()
    res = N * (N-1) // 2
    ans = [res]
    uf = UnionFind(N)
    for e in edge:
        a, b = e

        if uf.issame(a, b):
            ans.append(res)
        else:
            res -= uf.size(a)*uf.size(b)
            uf.unite(a, b)
            ans.append(res)
    
    ans.reverse()
    for i in range(1,M+1):
        print(ans[i])

if __name__=='__main__':
    main()